mypy/applytype.py

```diff
@@ -62,3 +68,32 @@ def apply_generic_arguments(callable: CallableType, types: List[Type],
         ret_type=expand_type(callable.ret_type, id_to_type),
         variables=remaining_tvars,
     )
+
+
+def get_incompatible_arg_constraints(arg_types: Sequence[Type], type: Type,
+                                     index: int) -> Dict[str, Tuple[str]]:
+    """Gets incompatible function arguments with the constrained types.
+
+    An example of a constrained type is AnyStr which must be all str or all byte.
+    """
+    constraints = {}  # type: Dict[str, Tuple[str]]
+    if isinstance(type, Instance) and type.type.name() == 'object':
+        if index == len(arg_types):
+            # Index is off by one for '*' arguments
+            constraints = add_arg_constraints(constraints, arg_types[index - 1])
+        else:
+            constraints = add_arg_constraints(constraints, arg_types[index])
+    return constraints
+
+
+def add_arg_constraints(constraints: Dict[str, Tuple[str]],
+                        arg_type: Type) -> Dict[str, Tuple[str]]:
+    if (isinstance(arg_type, TypeVarType) and
+            arg_type.values and
+            len(arg_type.values) > 1 and
+            arg_type.name not in constraints.keys()):
+        constraints[arg_type.name] = tuple(vals.type.name() for vals in arg_type.values)
+    elif isinstance(arg_type, UnionType):
+        for item in arg_type.items:
+            constraints = add_arg_constraints(constraints, item)
```

Comment:
The logic seems to be over-complicated here. You are basically just collecting a set of type variables affected by the error encountered. This should be separated from the error formatting logic (the latter should be kept in ``messages.py``).